AzureIoTHub.agent.lib.nut

```diff
@@ -732,44 +978,48 @@ class AzureIoTHub {
+
+            local devPath = "/" + _connStrParsed.DeviceId;
+            local username = format("%s%s/api-version=%s", _connStrParsed.HostName, devPath, AZURE_API_VERSION);
+            local resourcePath = format("/devices%s/api-version=%s", devPath, AZURE_API_VERSION);
+            _resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+            _mqttOptions = {
+                "username" : username,
+                "password" : null,
+                "keepalive" : _options.keepAlive
+            };
+            _url = "ssl://" + _connStrParsed.HostName;
         }
 
         // Opens a connection to Azure IoT Hub.
         //
         // Returns:                         Nothing.
         function connect() {
+            _log("Call: connect()");
             if (_isConnected || _isConnecting) {
-                _onConnectedCb && _onConnectedCb(_isConnected ? AZURE_IOT_CLIENT_ERROR_ALREADY_CONNECTED : AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                _onConnectedCb && _onConnectedCb(_isConnected ? AZURE_CLIENT_ERROR_ALREADY_CONNECTED : AZURE_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
                 return;
             }
-            _log("Connecting...");
-
-            local devPath = "/" + _connStrParsed.DeviceId;
-            local username = format("%s%s/api-version=%s", _connStrParsed.HostName, devPath, AZURE_API_VERSION);
-            local resourcePath = format("/devices%s/api-version=%s", devPath, AZURE_API_VERSION);
-            local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
-            local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
-            local sas = AzureIoTHub.SharedAccessSignature(resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
-
-            local options = {
-                "username" : username,
-                "password" : sas,
-                "keepalive" : _options.keepAlive
-            };
-
-            local url = "ssl://" + _connStrParsed.HostName;
 
+            _log("Connecting...");
             _isConnecting = true;
 
-            _mqttclient.connect(url, _connStrParsed.DeviceId, options);
+            _updatePasswd();
+            _mqttclient.connect(_url, _connStrParsed.DeviceId, _mqttOptions);
         }
 
         // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
         //
         // Returns:                         Nothing.
         function disconnect() {
+            _log("Call: disconnect()");
             if ((!_isDisconnected || _isConnecting) && !_isDisconnecting) {
                 _isDisconnecting = true;
+                if (_isRefreshingToken) {
+                    _log("Token refreshing is in progress now. Putting the request (disconnect) to the queue...");
+                    _pendingCalls.append(@() _mqttclient.disconnect(_onDisconnected.bindenv(this)));
```

Comment:
Don't we need to `bindenv(this)` the lambda function to let it access `_mqttclient `?